java/sage/io/RemoteSageFile.java

```diff
@@ -0,0 +1,489 @@
+    {
+      reconnect();
+      outStream.write(("READ " + remoteOffset + " " + len + "\r\n").getBytes(Sage.BYTE_CHARSET));
+      outStream.flush();
+      inStream.readFully(b, off, len);
+    }
+
+    remoteOffset += len;
+    return len;
+  }
+
+  @Override
+  public void readFully(byte[] b) throws IOException
+  {
+    readFully(b, 0, b.length);
+  }
+
+  @Override
+  public void readFully(byte[] b, int off, int len) throws IOException
+  {
+    if (len == 0)
+      return;
+
+    int bytesRead = read(b, off, len);
+
+    while (len > 0)
+    {
+      if (bytesRead == -1)
+        throw new EOFException("End of file reached!");
+
+      len -= bytesRead;
+      off += bytesRead;
+
+      if (len <= 0)
+        break;
+
+      bytesRead = read(b, off, len);
+    }
+  }
+
+  @Override
+  public synchronized void write(int b) throws IOException
+  {
+    if (readonly)
+      throw new IOException("Remote file is read only.");
+
+    try
+    {
+      outStream.write(("WRITE " + remoteOffset + " 1\r\n").getBytes(Sage.BYTE_CHARSET));
+      outStream.flush();
+      outStream.write(b);
+    }
+    catch (IOException e)
+    {
+      reconnect();
+      outStream.write(("WRITE " + remoteOffset + " 1\r\n").getBytes(Sage.BYTE_CHARSET));
+      outStream.flush();
+      outStream.write(b);
+    }
+
+    remoteOffset += 1;
+  }
+
+  @Override
+  public void write(byte[] b) throws IOException
+  {
+    write(b, 0, b.length);
+  }
+
+  @Override
+  public synchronized void write(byte[] b, int off, int len) throws IOException
+  {
+    if (readonly)
+      throw new IOException("Remote file is read only.");
+
+    if (len == 0)
+      return;
+
+    try
+    {
+      outStream.write(("WRITE " + remoteOffset + " " + len + "\r\n").getBytes(Sage.BYTE_CHARSET));
+      outStream.flush();
+      outStream.write(b, off, len);
+    }
+    catch (IOException e)
+    {
+      reconnect();
+      outStream.write(("WRITE " + remoteOffset + " " + len + "\r\n").getBytes(Sage.BYTE_CHARSET));
+      outStream.flush();
+      outStream.write(b, off, len);
+    }
+
+    remoteOffset += len;
+  }
+
+  @Override
+  public void randomWrite(long pos, byte[] b, int off, int len) throws IOException
+  {
+    if (readonly)
+      throw new IOException("Remote file is read only.");
+
+    long oldPosition = remoteOffset;
+    remoteOffset = pos;
+    try
+    {
+      write(b, off, len);
+    }
+    finally
+    {
+      remoteOffset = oldPosition;
+    }
+  }
+
+  @Override
+  public synchronized long skip(long n) throws IOException
+  {
+    if (n <= 0)
+      return 0;
+
+    long pos = remoteOffset;
+    long seek = Math.min(pos + n, length());
+
+    remoteOffset = seek;
+
+    return seek - pos;
+  }
+
+  @Override
+  public void seek(long pos)
+  {
+    remoteOffset = pos;
+  }
+
+  @Override
+  public long position()
+  {
+    return remoteOffset;
+  }
+
+  @Override
+  public long length() throws IOException
+  {
+    String response = executeCommand("SIZE\r\n");
```

Comment:
You can optimize this so that if the first time you send SIZE, you get back the same value for both, you never need to call the remote command again. Once a file goes non-active...it will never go active again.